### PR TITLE
Hide user UI

### LINKED ISSUE #48: Feature: Delete user UI
Delete a user using the UI as an admin. In the backend, this will mark the user as stale/inactive instead of removing them entirely. (Fix this to be a proper issue)

### LINKED ISSUE #84: Backend delete (hide) user
Add backend code to allow an admin to hide a user

### PR DESCRIPTION
**Pull request by publicqi**
Needs #85 to be merged first.

## Overview

In this PR, the frontend of the hide/unhide users is implemented. See https://ucsb-cs156-m23.github.io/proj-happycows-m23-10am-4/prs/86/storybook/?path=/story/components-users-userstable--three-users

It also has the changes to backend same as #85.

Here's a list of detailed changes for grading:
+ Backend: User entity has `isHidden` field and corresponding checks&api added
+ Backend: Tests adapted and added
+ Backend: New exception type `UserHiddenException`
+ Backend: Helper function to return a customized User in test (by `getUser`)
+ Frontend: Storybook page for UsersTables
+ Frontend: Add hide/unhide in UsersTables w/ tests

## Screenshots (Optional)

Hide/Unhide user page
image
Admin can hide/unhide users

image
A hidden user cannot do any operation (see backend checks). All endpoints with "ROLE_USER" are disabled.

## Future Possibilities (Optional)

Right now, all permission (hidden or not) checks are done at the endpoints. That means whenever a new api is added, the `user.isHidden()` needs to be checked. I wonder if there's a more elegant way (on top of everything, just like the `ROLE_*`) to do the checks.

## Tests

- [x] Backend Unit tests (`mvn test`) pass
- [x] Backend Test coverage (`mvn test jacoco:report`) 100%
- [x] Backend Mutation tests (`mvn test pitest:mutationCoverage`) 100% 
- [x] Frontend Unit tests (`npm test`) pass
- [x] Frontend Test coverage (`npm run coverage`) 100%
- [x] Frontend Mutation tests (`npx stryker run`) 100% 
- [x] Frontend Linting (`npx eslint --fix src`) 

## Linked Issues

Closes #84
Closes #48

### CODE REVIEW

**Review by iain801 — DISMISSED** [review]
Nice job! I like the way this is handled

**Comment by publicqi** [comment]
I'll fix the conflict when all other approved PRs get merged into main

File: frontend/src/main/components/Users/UsersTable.js [hunk 1]
```diff
@@ -1,33 +1,116 @@
...
+            accessor: (row, _rowIndex) => String(row.hidden) // hack needed for boolean values to show up
+        }
+    ];
+
+    /*
+    const columnsWithHide = [
+        ...columns,
+        ButtonColumn("Hide", "danger", deleteCallback, "UsersTable"),
+        ButtonColumn("Unhide", "success", unhideCallback, "UsersTable")
+    ];
```
Comment (pconrad) [inline frontend/src/main/components/Users/UsersTable.js]: Why is this commented out code here?

File: frontend/src/main/components/Users/UsersTable.js [hunk 2]
```diff
@@ -1,33 +1,116 @@
...
+        {
+            Header: 'Unhide',
+            id: 'unhide',
+            Cell: ({ cell }) => (
+                <Button
+                    variant={"success"}
+                    onClick={() => unhideCallback(cell)}
+                    data-testid={`UsersTable-cell-row-${cell.row.index}-col-${cell.column.id}-button`}
+                    disabled={cell.row.values["hidden"] === "false" ||
+                    // Stryker disable all : backend does not support admin hiding
```
Comment (pconrad) [inline frontend/src/main/components/Users/UsersTable.js]: I don't understand the comment "backend does not support admin hiding".   I mean, I understand the comment, but that's not a reason to exclude this from Stryker mutation testing. In fact, this should most certainly be tested with a frontend tests (one that has an admin user and asserts that the button is disabled, and another than is a non-admin users and asserts that the button is not disabled.)

File: frontend/src/stories/components/Users/UsersTable.stories.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    currentUser: currentUserFixtures.adminUser
+};
+
+export const OneCommonsAdmin = Template.bind({});
+
+OneCommonsAdmin.args = {
+    commons: commonsPlusFixtures.oneCommonsPlus,
+    currentUser: currentUserFixtures.adminUser
+}
+*/
```
Comment (pconrad) [inline frontend/src/stories/components/Users/UsersTable.stories.js]: commented out code?

**Review by pconrad — CHANGES_REQUESTED** [review]
There a few things I'm not sure about

File: frontend/src/main/components/Users/UsersTable.js (hunk 1 above)
Comment (publicqi) [reply frontend/src/main/components/Users/UsersTable.js]: This is a leftover code snippet. Initially I want to create hide/unhide button for every user and they are all enabled. Later I wrote the version that will disable buttons based on the "hidden" and "admin"attribute. I will remove this.

File: frontend/src/stories/components/Users/UsersTable.stories.js (hunk 3 above)
Comment (publicqi) [reply frontend/src/stories/components/Users/UsersTable.stories.js]: Again this is leftover code copied from Commons story. I will remove this.

File: frontend/src/main/components/Users/UsersTable.js (hunk 2 above)
Comment (publicqi) [reply frontend/src/main/components/Users/UsersTable.js]: I'm a little confused about this one. The code implies that the backend will always return a user with
1. admin and not hidden
2. user and hidden
3. user and not hidden

So a hidden admin is not considered. Should I just include a test with hidden admin?

```js
    test("admin will always disable both buttons", () => {
        const testId = "UsersTable";
        let baseUser = usersFixtures.threeUsers[0];

        baseUser.admin = true;
        baseUser.hidden = true;
        render(
            <QueryClientProvider client={queryClient}>
                <UsersTable users={[baseUser]}/>
            </QueryClientProvider>
        );

        expect(screen.getByTestId(`${testId}-cell-row-0-col-hide-button`)).toBeDisabled();
        expect(screen.getByTestId(`${testId}-cell-row-0-col-unhide-button`)).toBeDisabled();
      });
```

File: frontend/src/main/components/Users/UsersTable.js (hunk 1 above)
Comment (publicqi) [reply frontend/src/main/components/Users/UsersTable.js]: https://github.com/ucsb-cs156-m23/proj-happycows-m23-10am-4/pull/86/commits/79239c3485bc4b8a3cd5c2754f9dbc573cc9e8d1

File: frontend/src/stories/components/Users/UsersTable.stories.js (hunk 3 above)
Comment (publicqi) [reply frontend/src/stories/components/Users/UsersTable.stories.js]: https://github.com/ucsb-cs156-m23/proj-happycows-m23-10am-4/pull/86/commits/79239c3485bc4b8a3cd5c2754f9dbc573cc9e8d1

**Comment by publicqi** [comment]
This PR has been replaced by #101

**Pull request closed without merging**